PYTHON/stencil_ray.py

```diff
@@ -0,0 +1,369 @@
+        if (2 * r + 1) > n:
+            sys.exit("ERROR: Stencil radius exceeds grid size")
+    else:
+        r = 2  # radius=2 is what other impls use right now
+
+    print('Grid size            = ', n)
+    print('Radius of stencil    = ', r)
+    print('Type of stencil      = ', 'star')
+    print('Data type            = double precision')
+    print('Compact representation of stencil loop body')
+    print('Number of iterations = ', iterations)
+    print('Number of processes  = ', num_procs)
+
+    ray.init(address='auto', node_ip_address='sr231', redis_password='123')
+
+    num_procs_x, num_procs_y = factor(num_procs)
+
+    executors = []
+    for i in range(num_procs):
+        id_x = i % num_procs_x
+        id_y = i // num_procs_x
+
+        i_start, i_end = compute_star_end(id_x, n, num_procs_x)
+        width = i_end - i_start + 1
+        if width == 0:
+            print(f"ERROR: rank {i} has no work to do")
+            sys.exit(1)
+
+        j_start, j_end = compute_star_end(id_y, n, num_procs_y)
+        height = j_end - j_start + 1
+        if height == 0:
+            print(f"ERROR: rank {i} has no work to do")
+            sys.exit(1)
+
+        if width < r or height < r:
+            print(f"ERROR: rank {i} has work tile smaller then stencil radius")
+            sys.exit(1)
+
+        executors.append(Executor.remote(n, r, i, num_procs, num_procs_x, num_procs_y,
+                                         i_start, i_end, j_start, j_end))
+
+    for i in range(num_procs):
+        id_x = i % num_procs_x
+        id_y = i // num_procs_x
+
+        left = executors[i - 1] if id_x > 0 else None
+        right = executors[i + 1] if id_x < num_procs_x - 1 else None
+        top = executors[i + num_procs_x] if id_y < num_procs_y - 1 else None
+        bottom = executors[i - num_procs_x] if id_y > 0 else None
+
+        ray.get(executors[i].register.remote(left, right, top, bottom))
+
+    for k in range(iterations + 1):
+        # start timer after a warmup iteration
+        if k == 1:
+            t0 = time.time()
+
+        idss = ray.get([executors[i].execute.remote() for i in range(num_procs)])
+        for ids in idss:
+            ray.get(ids)
+
+        ray.get([executors[i].apply_stencil_operator.remote() for i in range(num_procs)])
+
+    t1 = time.time()
+    stencil_time = t1 - t0
+
+    # ******************************************************************************
+    # * Analyze and output results.
+    # ******************************************************************************
+
+    abserrors = [ray.get(executors[i].abserror.remote()) for i in range(num_procs)]
+    norm = sum(abserrors)
+    active_points = (n - 2 * r) ** 2
+    norm /= active_points
+
+    epsilon = 1.e-8
+
+    ray.shutdown()
+
+    # verify correctness
+    reference_norm = 2 * (iterations + 1)
+    if abs(norm - reference_norm) < epsilon:
+        print('Solution validates')
+        flops = (2 * r + 1) * active_points
```

Comment:
It seems this should be (4 * r + 1) * active_points ?

Reply:
This keeps the same as the other stencil implementation.

Reply:
I saw below in PYTHON/stencil.py
stencil_size = 4 * r + 1
flops = (2 * stencil_size + 1) * active_points

Reply:
You are right. I will correct it.